### PR TITLE
[Refactor] Refactor materialized view partition synchronize logical

### LINKED ISSUE #8622: Materialize view cause FE to fail to start
### Steps to reproduce the behavior (Required)
CREATE TABLE `fact` (
  `datekey` int(11) NULL COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(`datekey`)
COMMENT "OLAP"
PARTITION BY RANGE(`datekey`)
(
PARTITION p20220627 VALUES [("20220627"), ("20220628")),
PARTITION p20220628 VALUES [("20220628"), ("20220629"))
)
DISTRIBUTED BY HASH(`datekey`) BUCKETS 1
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
);

insert into fact values('20220627');
insert into fact values('20220628');
insert into dim values('20220627');
insert into dim values('20220628');

CREATE TABLE `dim` (
  `product_key` varchar(100) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`product_key`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`product_key`) BUCKETS 1
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
);

CREATE MATERIALIZED VIEW test_mv
COMMENT "test_MV"
PARTITION BY datekey
DISTRIBUTED BY HASH(datekey)
REFRESH ASYNC AS
SELECT a.datekey
  FROM fact a
  join dim b
    on a.datekey = b.product_key

ALTER TABLE fact drop PARTITION p20220628

when checkpoint do following
refresh materialized view test_mv

restart fe

### Expected behavior (Required)
fe start success.
### Real behavior (Required)
image

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
2.4

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #8622

## Problem Summary(Required) ：

1. Separate the logic, first calculate the partitions that need to be added and deleted, and then perform the corresponding actions.
2. Remove tableMvPartitionNameRefMap & mvTablePartitionNameRefMap use rule-based partition mapping.
3. Use the new algorithm to calculate the corresponding partition
4. Consider the issue of flush propagation

### CODE REVIEW

**Comment by Astralidea** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/scheduler/MvTaskRunProcessor.java [hunk 1]
```diff
@@ -169,89 +167,54 @@ public void processTaskRun(TaskRunContext context) {
...
+        Map<String, Range<PartitionKey>> basePartitionMap = base.getRangePartitionMap();
+        Map<String, Range<PartitionKey>> mvPartitionMap = mv.getRangePartitionMap();
+
+        PartitionDiff partitionDiff = new PartitionDiff(Maps.newHashMap(), Maps.newHashMap());
+        if (partitionExpr instanceof SlotRef) {
+            partitionDiff = ExpressionPartitionUtil.calcSyncSamePartition(basePartitionMap, mvPartitionMap);
+        } else if (partitionExpr instanceof FunctionCallExpr) {
+            FunctionCallExpr functionCallExpr = (FunctionCallExpr) partitionExpr;
+            String functionName = functionCallExpr.getFnName().getFunction();
+            if (!functionName.equalsIgnoreCase("date_trunc")) {
```
Comment (blackstar-baba) [inline fe/fe-core/src/main/java/com/starrocks/scheduler/MvTaskRunProcessor.java:182]: `date_trunc` is checked when creating mv

File: fe/fe-core/src/main/java/com/starrocks/sql/common/ExpressionPartitionUtil.java [hunk 2]
```diff
@@ -1,205 +1,224 @@
...
-            LiteralExpr upperLiteralExpr = ExpressionPartitionUtil.getBound(
-                    functionCallExpr, baseUpperLiteralExpr, partitionColumn.getPrimitiveType(), 0);
-            if (ExpressionPartitionUtil.compareBound(baseUpperLiteralExpr, upperLiteralExpr) > 0) {
-                // increase upperBound
-                upperLiteralExpr = ExpressionPartitionUtil.getBound(
-                        functionCallExpr, upperLiteralExpr, partitionColumn.getPrimitiveType(), 1);
-            }
+    public static Map<String, Range<PartitionKey>> mappingRangeList(Map<String, Range<PartitionKey>> baseRangeMap,
+                                                              String granularity) {
+        Set<LocalDateTime> timePointSet = Sets.newTreeSet();
```
Comment (blackstar-baba) [inline fe/fe-core/src/main/java/com/starrocks/sql/common/ExpressionPartitionUtil.java:59]: Should we be limited to time types ?

File: fe/fe-core/src/main/java/com/starrocks/scheduler/MvTaskRunProcessor.java [hunk 3]
```diff
@@ -169,89 +165,52 @@ public void processTaskRun(TaskRunContext context) {
...
-                    dropPartition(database, materializedView, refMvPartitionName);
-                } else {
-                    needRefreshPartitionNames.add(refMvPartitionName);
-                }
-            }
-            materializedView.removeBasePartition(baseTableId, deletedPartitionName);
+
+        Map<String, Range<PartitionKey>> adds = partitionDiff.getAdds();
+        Map<String, Range<PartitionKey>> deletes = partitionDiff.getDeletes();
+        Set<String> needRefreshPartitionNames = Sets.newHashSet();
```
Comment (blackstar-baba) [inline fe/fe-core/src/main/java/com/starrocks/scheduler/MvTaskRunProcessor.java:192]: Do we have partitions that need to be refreshed directly?

File: fe/fe-core/src/main/java/com/starrocks/sql/common/ExpressionPartitionUtil.java (hunk 2 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/sql/common/ExpressionPartitionUtil.java]: I think DateTime for rollup will be enough

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java [hunk 4]
```diff
@@ -650,6 +650,18 @@ public PartitionInfo getPartitionInfo() {
         return partitionInfo;
     }
 
+    // partition Name -> Range
+    public Map<String, Range<PartitionKey>> getRangePartitionMap() {
+        Preconditions.checkArgument(partitionInfo.getType() == PartitionType.RANGE);
```
[the remark this answers is not in the document]
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:656]: UNPARTITIONED will not call here I have already tested.

File: fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    public static void calcPotentialRefreshPartition(Set<String> needRefreshMvPartitionNames,
+                                                      Set<String> baseChangedPartitionNames,
+                                                      Map<String, Set<String>> baseToMvNameRef,
+                                                      Map<String, Set<String>> mvToBaseNameRef) {
+        int curNameCount = needRefreshMvPartitionNames.size();
+        int updatedCount = SyncPartitionUtils.gatherPotentialRefreshPartitionNames(baseChangedPartitionNames,
+                needRefreshMvPartitionNames, baseToMvNameRef, mvToBaseNameRef);
+        while (curNameCount != updatedCount) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java:138]: better to use do while

**Review by ABingHuang — COMMENTED** [review]
LGTM

File: fe/fe-core/src/main/java/com/starrocks/sql/common/ExpressionPartitionUtil.java (hunk 2 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/sql/common/ExpressionPartitionUtil.java]: you are right. we should also support type DATE

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 267 / 300 (89.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/PartitionDiff.java|13|18|72.22%|[29, 41, 42, 49, 50]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|143|164|87.20%|[35, 73, 90, 91, 145, 146, 169, 170, 172, 173, 181, 182, 184, 185, 187, 198, 205, 227, 236, 242, 270]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunProcessor.java|64|70|91.43%|[122, 123, 244, 247, 263, 264]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|11|12|91.67%|[20]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/DateUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/PartitionMapping.java|6|6|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 263 / 302 (87.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/PartitionDiff.java|13|18|72.22%|[29, 41, 42, 49, 50]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunProcessor.java|59|70|84.29%|[122, 123, 208, 240, 241, 243, 244, 246, 247, 263, 264]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|143|165|86.67%|[44, 45, 85, 102, 103, 157, 158, 181, 182, 184, 185, 193, 194, 196, 197, 199, 210, 217, 239, 248, 254, 282]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|11|12|91.67%|[20]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/DateUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/PartitionMapping.java|6|6|100.00%|[]|

**Review by ABingHuang — APPROVED** [review]
LGTM

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by chaoyli — APPROVED** [review]

**Pull request merged**